### ISSUE TITLE
Add support for OAuth 2.0 for newsletter agents

### ISSUE THREAD

**Issue by johnsonwalker** [JonnyWong16/plexpy#2407]
### Is your feature request related to a problem?

Outlook.com no longer supports 'basic authentication' - see Microsoft article: [Modern Authentication Methods now needed to continue syncing Outlook Email in non-Microsoft email apps](https://support.microsoft.com/en-au/office/modern-authentication-methods-now-needed-to-continue-syncing-outlook-email-in-non-microsoft-email-apps-c5d65390-9676-4763-b41f-d7986499a90d).

This has broken my newsletter agent which was using my Hotmail email address to send from.

### What is your feature request?

Add OAuth 2.0 support to allow newsletter agents using Outlook/Microsoft email services for sending to continue to function.

### Are there any workarounds?

Create a new email with a different provider who still supports basic auth (not sure who - and most services will likely force OAuth2 at some point).

**Comment by JonnyWong16** [comment]
You can use an app password.

https://support.microsoft.com/en-us/account-billing/how-to-get-and-use-app-passwords-5896ed9b-4263-e681-128a-a6f2979a7944

**Comment by johnsonwalker** [comment]
> You can use an app password.
> 
> https://support.microsoft.com/en-us/account-billing/how-to-get-and-use-app-passwords-5896ed9b-4263-e681-128a-a6f2979a7944

Not anymore. I’ve been using an app password this whole time (with two step verification enabled on my account) and it only stopped working this past week. The error in the log being `ERROR   :: Thread-2 (process_queue) : Tautulli Notifiers :: Email notification failed: (535, b'5.7.139 Authentication unsuccessful, basic authentication is disabled.` If you read the article I posted, or this one [here](https://learn.microsoft.com/en-us/exchange/clients-and-mobile-in-exchange-online/deprecation-of-basic-authentication-exchange-online), you’ll see that Microsoft have stated that _The deprecation of basic authentication also prevents the use of app passwords with apps that don't support two-step verification._

**Comment by JonnyWong16** [comment]
I just tested an app password with my own account this morning before I replied. 

Before using an app password:
```
2024-09-22 07:49:18 - DEBUG   :: CP Server Thread-6 : Sending test Email notification.
2024-09-22 07:49:18 - DEBUG   :: CP Server Thread-6 : Tautulli NotificationHandler :: Notifiers enabled for notify_action 'test' (manual trigger).
2024-09-22 07:49:18 - INFO    :: Thread-3 (process_queue) : Tautulli NotificationHandler :: Preparing notification for notifier_id 19.
2024-09-22 07:49:24 - ERROR   :: Thread-3 (process_queue) : Tautulli Notifiers :: Email notification failed: (535, b'5.7.139 Authentication unsuccessful, the user credentials were incorrect. [MW4P221CA0008.NAMP221.PROD.OUTLOOK.COM 2024-09-22T14:49:24.450Z 08DCD5DA5BAA8F77]')
```
After using an app password:
```
2024-09-22 07:49:50 - INFO    :: CP Server Thread-13 : Tautulli Notifiers :: Updated notification agent: Email (notifier_id 19).
2024-09-22 07:49:50 - DEBUG   :: CP Server Thread-14 : Sending test Email notification.
2024-09-22 07:49:50 - DEBUG   :: CP Server Thread-14 : Tautulli NotificationHandler :: Notifiers enabled for notify_action 'test' (manual trigger).
2024-09-22 07:49:50 - INFO    :: Thread-2 (process_queue) : Tautulli NotificationHandler :: Preparing notification for notifier_id 19.
2024-09-22 07:49:57 - INFO    :: Thread-2 (process_queue) : Tautulli Notifiers :: Email notification sent.
```

**Comment by johnsonwalker** [comment]
what smtp server are you using?

**Comment by JonnyWong16** [comment]
The one in the Wiki.

https://github.com/Tautulli/Tautulli/wiki/Notification-Agents-Guide#outlookcom

**Comment by johnsonwalker** [comment]
yeah that's the one i'm using. just created a new test one and still getting the same error:
```
2024-09-23 07:52:57 - DEBUG   :: CP Server Thread-10 : Sending test Recently Added newsletter.
2024-09-23 07:52:57 - INFO    :: Thread-3 (process_queue) : Tautulli NewsletterHandler :: Preparing newsletter for newsletter_id 3.
2024-09-23 07:52:59 - DEBUG   :: Thread-3 (process_queue) : Tautulli Helpers :: Image 'Upright' (poster) uploaded to Imgur.
2024-09-23 07:53:01 - DEBUG   :: Thread-3 (process_queue) : Tautulli Helpers :: Image 'Upright' (art) uploaded to Imgur.
2024-09-23 07:53:01 - INFO    :: Thread-3 (process_queue) : Tautulli Newsletters :: Generating newsletter.
2024-09-23 07:53:01 - INFO    :: Thread-3 (process_queue) : Tautulli Newsletters :: Recently Added newsletter saved to 'newsletter_ecc7eb7e.html'
2024-09-23 07:53:06 - ERROR   :: Thread-3 (process_queue) : Tautulli Notifiers :: Email notification failed: (535, b'5.7.139 Authentication unsuccessful, basic authentication is disabled. [MEAPR01CA0028.ausprd01.prod.outlook.com 2024-09-22T21:53:07.147Z 08DCDB4BB098F296]')
```

**Comment by JonnyWong16** [comment]
I'm not using an Exchange email. I'm just using a regular email from outlook.com.

**Comment by johnsonwalker** [comment]
Same. I've been using my @hotmail.com email and just tried my other @outlook.com email with the same result:

```
2024-09-23 13:15:51 - DEBUG   :: CP Server Thread-8 : Sending test Recently Added newsletter.
2024-09-23 13:15:51 - INFO    :: Thread-2 (process_queue) : Tautulli NewsletterHandler :: Preparing newsletter for newsletter_id 3.
2024-09-23 13:15:51 - INFO    :: Thread-2 (process_queue) : Tautulli Newsletters :: Generating newsletter.
2024-09-23 13:15:51 - INFO    :: Thread-2 (process_queue) : Tautulli Newsletters :: Recently Added newsletter saved to 'newsletter_d5cc0f80.html'
2024-09-23 13:15:56 - ERROR   :: Thread-2 (process_queue) : Tautulli Notifiers :: Email notification failed: (535, b'5.7.139 Authentication unsuccessful, basic authentication is disabled. [ME3P282CA0144.AUSP282.PROD.OUTLOOK.COM 2024-09-23T03:15:57.724Z 08DCDB450B179655]')
```

**Comment by johnsonwalker** [comment]
Found this [reddit post](https://www.reddit.com/r/MicrosoftOutlook/comments/1fk28xk/outlook_personal_accounts_security_changes_alert/) - looks like it's a staggered deprecation of basic auth. My region must be one of the first to be hit.

Also found this Microsoft [tech community post](https://techcommunity.microsoft.com/t5/outlook-blog/keeping-our-outlook-personal-email-users-safe-reinforcing-our/ba-p/4164184) and [comment](https://techcommunity.microsoft.com/t5/outlook-blog/keeping-our-outlook-personal-email-users-safe-reinforcing-our/bc-p/4165630/highlight/true#M5184) from the author (Microsoft employee) answering someone's question about app passwords.

![image](https://github.com/user-attachments/assets/d0aa3904-6348-42e1-8409-032f7b11b7ad)

**Comment by alexkahler** [comment]
I have the same issue. None of my email services work - from Overseerr to NAS notifications :/

**Comment by tw333k** [comment]
This is now happening to me using smtp-mail.outlook.com to send emails
`ERROR	Tautulli Notifiers :: Email notification failed: (535, b'5.7.139 Authentication unsuccessful, basic authentication is disabled. [SY5PR01CA0055.ausprd01.prod.outlook.com 2024-10-04T07:08:34.956Z 08DCE394E3F95462]')`

**Comment by Ja50n-Kru** [comment]
Yep, my account has now been hit too - no longer allows basic authentication. 

Is this being ignored / not possible / not interested - be good to know if it is on the radar.

**Comment by ClearlyDazed** [comment]
Microsoft email server is requiring 3rd party apps to have Oauth2 authorization. Having all 3rd party apps to compile with this would be a daunting task. To get past this, an email oauth2 proxy can be used as found here ( https://github.com/simonrob/email-oauth2-proxy ) and for docker found here (https://github.com/blacktirion/email-oauth2-proxy-docker ). Microsoft Azure account is needed to register the email proxy. Follow the associated instructions and hints found in  the issues open and closed sections.

In my case both the emailproxy and tautulli are in separate dockers on a linux server. The configuration of the tautulli email client specifies the SMTP server as emailproxy with port 1587 with the usual username and password and no encryption. For tautulli docker, I also connected the docker network for emailproxy.

**Comment by JamminR** [comment]
Gmail will soon be phasing out basic authentication also.
https://support.google.com/a/answer/14114704?hl=en

**Comment by xenago** [comment]
> Gmail will soon be phasing out basic authentication also. https://support.google.com/a/answer/14114704?hl=en

From that page it seems like they explicitly support app passwords, so I don't think this will be the same problem.

> For scanners or other devices using SMTP or less secure apps to send emails, use one of the following options:
> (...)
> Configure an app password for use with the device.

**Comment by JamminR** [comment]
> From that page it seems like they explicitly support app passwords, so I don't think this will be the same problem.
> Configure an app password for use with the device.

Personally, already am. Gmail sent some notice to the account I use 2-3 years ago.
I'm posting here in hopes OAuth will move farther up the priority list, as most if not all mail services are moving away from basic authentication. I imagine with the way cloud services seemingly arbitrarily change rules on a whim, some day, with little notice, that 'exception' for app passwords goes away.

**Comment by JonnyWong16** [comment]
Just FYI, self-hosted OAuth requires _a lot_ of setup from the user's end (i.e. your end). It's not as simple as just entering a few settings into Tautulli. 

This is an example for Microsoft 365. Everything up until the bit of python code at the bottom (Azure application and permissions) must be set up by you.

https://medium.com/preprintblog/how-to-send-emails-using-microsoft-365-in-python-flask-with-oauth2-21becefd2d64

It's a very big ask for Tautulli users to go through all these steps.

**Comment by JamminR** [comment]
I admit - in my mind i may be over simplifying what I remember I had to do over past 4 years for two hardware devices (A rack UPS and a Synology DSM), but in some way they had me input a token, or they went and got token after I input my Gmail account login once. I occasionally have to refresh them for them to work by getting a new token (by refresh/relogging in).
I would hope that with as simple as the setup of my other devices 'seemed', Tautulli could do the same. (I thought the Tautulli>Plex login was Oauth, but could be wrong)

**Comment by xenago** [comment]
Yeah, from the user side it's supposed to be easy. Every email client supports oauth now, including open source ones, and users never need to do anything fancy. So it's absolutely doable. Thunderbird etc all do it.

**Comment by JonnyWong16** [comment]
That's if I create and host the application. I have no interest in doing that.

Also because Tautulli is open source, I cannot embed the secret keys into the source code.

**Comment by JamminR** [comment]
Understood. Perhaps Tautulli could integrate a 3rd party Oauth provider like it does for images/Cloudinary?
I would presume similar for 3rd party apps exists. 
Just a quick duckduckgo search resulted in what seems to be a "oauth for 3rd party" app - pricing is 3 free connections, any more than that requires payment.
https://www.nango.dev/auth
I may be missunderstanding though. If nango is considering each download of Tautulli as a connection, and not MY instance of Tautulli as a single on it's own, well, i can see that cost being way too high.

Either way, we/I for now have gone beyond full scope of this request - 
summarized - it would be nice if we could get some free way to integrate Oauth into Tautulli, be it directly or 3rd party. Let's Encrypt came out with free SSL certs, maybe a Oauth service will come to light also.
I do get a sense that email services will be moving away from even app passwords.

**Comment by JonnyWong16** [comment]
A unified API like nango doesn't solve the problem.

If I am the one that sets up nango, then it's the same problem - I cannot embed my nango secret keys into Tautulli's source code. Plus I would have also have access to everyone's Google/Microsoft account tokens - something I definitely don't want to be responsible for.

If I ask the user to setup nango themselves, then it's also the same problem with asking the user to do some complicated setup.

All these tools are design for the user to be the client side, and the developer to control the server side. But because Tautulli is self-hosted, the user is both the client side and server side. I do not want to be the server side for all Tautulli users.
